backend/icpy/services/preview_service.py

```
@@ -0,0 +1,894 @@
+        raise Exception("No available ports in range")
+
+    def _release_port(self, port: int):
+        """Release a port back to the available pool."""
+        self.used_ports.discard(port)
+
+    def _generate_content_hash(self, files: Dict[str, str], project_type: str) -> str:
+        """Generate a hash for the files content to identify identical previews."""
+        # Sort files by name for consistent hashing
+        sorted_files = sorted(files.items())
+        content_str = f"{project_type}:" + ":".join(f"{name}:{content}" for name, content in sorted_files)
+        hash_value = hashlib.md5(content_str.encode()).hexdigest()
```

Ruff (0.12.2)

Probable use of insecure hash functions in `hashlib`: `md5`

(S324)